File: 213-house-robber-ii/213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        
        ans = 0
        for start, m in [(0, n - 1), (1, n)]:
            prev = curr = 0
            for pos in range(start, m):
                prev, curr = curr, max(nums[pos] + prev, curr)
            ans = max(ans, curr)
        return ans

File: 213-house-robber-ii/213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        if(nums.length == 1)
            return nums[0];
            
        int[][] params = new int[][]{
            new int[]{0, nums.length - 1}, 
            new int[]{1, nums.length}
        };
        
        int ans = 0;
        for(int[] param : params){
            int start = param[0], m = param[1];
            int prev = 0, curr = 0;
            for(int i = m - 1; i >= start; i--){
                int tmp = prev;
                prev = curr;
                curr = Math.max(nums[i] + tmp, curr);
            }
            ans = Math.max(ans, curr);
        }
        return ans;
    }
}